File: client/src/com/networkapi/CircularBuffer.java

package com.networkapi;

import java.util.concurrent.Semaphore;

public class CircularBuffer {

    private byte[] buffer;

    private int head, tail, lenght;

    private boolean bufferFull;

    private Semaphore sem;

    public CircularBuffer(int size) {
        this.lenght = size;

        buffer = new byte[size];

        head = tail = 0;

        bufferFull = false;

        sem = new Semaphore(1);
    }

    public void write(byte[] values) throws InterruptedException {
        write(values, values.length);
    }

    public void write(byte[] values, int size) throws InterruptedException {
        sem.acquire();

        for (int i = 0; i < size; i++) {
            buffer[tail++] = values[i];

            tail = tail % lenght;

            if (tail > head)
                bufferFull = (tail - head) == size;
            else
                bufferFull = (tail + lenght - head) == size;

            if (bufferFull == true)
                break;
        }

        sem.release();
    }

    public byte[] read() throws InterruptedException {

        sem.acquire();

        byte ret[];

        if (tail > head)
            ret = new byte[tail - head];
        else
            ret = new byte[tail + lenght - head];

        for (int i = 0; i < ret.length; i++) {
            ret[i] = buffer[head++];

            head = head % lenght;
        }

        sem.release();

        return ret;
    }

    /**
     * @brief Just read values from circular buffer. Don't change head or tail
     *        values.
     * @throws InterruptedException
     */
    public byte[] get() throws InterruptedException {
        sem.acquire();

        byte ret[];
        int headCopy = head;
        int iSize;

        if (tail == head) {
            iSize = 0;
        }
        else if (tail > head) {
            iSize = tail - head;
        }
        else {
            iSize = tail + lenght - head;
        }

        ret = new byte[iSize];

        for (int i = 0; i < ret.length; i++) {
            ret[i] = buffer[headCopy++];
            headCopy = headCopy % lenght;
        }

        sem.release();

        return ret;
    }

    public void erase(int howMuch) throws InterruptedException {
        sem.acquire();

        head += howMuch;
        head = head % lenght;

        sem.release();
    }

    public boolean isFull() {
        return bufferFull;
    }
}

File: server/CircularBuffer.cpp
#include "CircularBuffer.h"

#include <iostream>
#include <stdio.h>

using namespace std;

pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

CircularBuffer::CircularBuffer(unsigned int _uiSize) :
    uiLength(_uiSize) {
    pucBuffer = new unsigned char[uiLength];

    uiHead = uiTail = 0;

    bBufferFull = false;
}

CircularBuffer::~CircularBuffer() {
    if (pucBuffer != NULL) {
        delete pucBuffer;
        pucBuffer = NULL;
    }
}

void CircularBuffer::write(unsigned char * pucValues, unsigned int uiSize) {
    pthread_mutex_lock(&mutex);

    cout << endl << "CB write: head = " << uiHead << ", tail = " << uiTail << ", size = " << uiSize;
    cout << endl << "[ ";

    for (unsigned int i = 0; i < uiSize; i++) {
        pucBuffer[uiTail++] = pucValues[i];
        printf("0x%02x ", pucValues[i]);
        uiTail = uiTail % uiLength;

        if (uiTail > uiHead)
            bBufferFull = (uiTail - uiHead) == uiLength;
        else
            bBufferFull = (uiTail + uiSize - uiHead) == uiLength;

        if (bBufferFull == true)
            break;
    }

    cout << "]";

    pthread_mutex_unlock(&mutex);
}

unsigned char * CircularBuffer::read() {

    pthread_mutex_lock(&mutex);

    unsigned char * pucRet;

    unsigned int uiSize = uiTail > uiHead ? uiTail - uiHead : uiTail + uiLength - uiHead;
    cout << endl << "CB read: head = " << uiHead << ", tail = " << uiTail << ", size = " << uiSize;
    cout << endl << "[ ";

    pucRet = new unsigned char[uiSize];

    for (unsigned int i = 0; i < uiSize; i++) {
        pucRet[i] = pucBuffer[uiHead++];
        printf("0x%02x ", pucRet[i]);
        uiHead = uiHead % uiLength;
    }

    cout << "]";

    pthread_mutex_unlock(&mutex);

    return pucRet;
}

/**
 * @brief Just read values from circular buffer. Don't change head or tail values.
 */
unsigned char * CircularBuffer::get(unsigned int * puiLength) {
    pthread_mutex_lock(&mutex);

    unsigned char * pucRet;

    unsigned int uieadCopy = uiHead;

    if (uiTail == uiHead) {
        *puiLength = 0;
    } else if (uiTail > uiHead) {
        *puiLength = uiTail - uiHead;
    } else {
        *puiLength = uiTail + uiLength - uiHead;
    }
    cout << endl << "CB get: head = " << uiHead << ", tail = " << uiTail << ", size = "
        << *puiLength;
    pucRet = *puiLength == 0 ? NULL : new unsigned char[*puiLength];
    cout << endl << "[ ";
    for (unsigned int i = 0; i < *puiLength; i++) {
        pucRet[i] = pucBuffer[uieadCopy++];
        printf("0x%02x ", pucRet[i]);
        uieadCopy = uieadCopy % uiLength;
    }
    cout << "]";
    pthread_mutex_unlock(&mutex);

    return pucRet;
}

void CircularBuffer::erase(unsigned int uiHowMuch) {
    pthread_mutex_lock(&mutex);

    cout << endl << "CB erase: head = " << uiHead << ", tail = " << uiTail << ", how much = "
        << uiHowMuch;

    uiHead += uiHowMuch;
    uiHead = uiHead % uiLength;

    pthread_mutex_unlock(&mutex);
}

bool CircularBuffer::isFull() {
    return bBufferFull;
}
